compiler/src/iree/compiler/Dialect/Stream/Transforms/FoldUniformOperands.cpp

```diff
@@ -252,6 +273,115 @@ inlineUniformConstants(mlir::FunctionOpInterface funcOp,
+  for (auto [i, dispatchOp] : llvm::enumerate(dispatchOps)) {
+    SmallVector<unsigned> &dupeIndexMap = dupeIndexMaps[i];
+    dupeIndexMap.resize(workloadCount, kUnassigned);
+    SmallVector<Value> workloads = dispatchOp.getWorkload();
+    for (unsigned i = 0; i < workloads.size(); ++i) {
+      for (unsigned j = 0; j < i; ++j) {
+        if (workloads[j] == workloads[i]) {
+          dupeIndexMap[i] = j;
+          break;
+        }
+      }
+    }
+  }
+
+  // Per-workload now find which are consistently duplicated.
+  auto [uniformDupeIndexMap, deadOperandsMap] =
+      getUniformDupeIndexingAndDeadValues(dupeIndexMaps);
+  if (deadOperandsMap.none()) {
+    // No-op.
+    return;
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "deduplicateWorkloads for " << exportOp.getName() << "\n";
+    llvm::dbgs() << "  dead workloads: ";
+    llvm::interleaveComma(deadOperandsMap.set_bits(), llvm::dbgs());
+    llvm::dbgs() << "\n";
+    for (auto [deadIdx, liveIdx] : llvm::enumerate(uniformDupeIndexMap)) {
+      if (deadIdx == liveIdx) {
+        continue;
+      }
+      llvm::dbgs() << "  %arg" << deadIdx << " -> %arg" << liveIdx << "\n";
+    }
+  });
+
+  // Update the ordinals in the workload ordinal ops.
+  unsigned ordinalCount = 0;
+  SmallVector<unsigned> newOrdinals(uniformDupeIndexMap.size());
+  for (auto [deadIdx, liveIdx] : llvm::enumerate(uniformDupeIndexMap)) {
+    if (deadIdx > liveIdx) {
+      continue;
+    }
+    newOrdinals[deadIdx] = ordinalCount++;
+  }
+  mlir::FunctionOpInterface funcOp = exportOp.lookupFunctionRef();
+  funcOp->walk([&](IREE::TensorExt::DispatchWorkloadOrdinalOp ordinalOp) {
+    uint64_t ordinal = ordinalOp.getOrdinal().getZExtValue();
+    assert(ordinal < newOrdinals.size() &&
+           "the ordinal shouldn't be larger than the number of workloads");
+    uint64_t newOrdinal = newOrdinals[ordinal];
+    if (newOrdinal == ordinal) {
+      return WalkResult::advance();
+    }
+    ordinalOp.setOrdinal(APInt(64, newOrdinal));
+    return WalkResult::advance();
+  });
+
+  // Update the workgroup region arguments
+  Block &entryBlock = exportOp.getWorkgroupCount().front();
+  auto argReplacementMap = llvm::to_vector(
+      llvm::seq(0u, entryBlock.getNumArguments())); // old -> new
+  llvm::BitVector deadArgMap(entryBlock.getNumArguments());
+  for (auto [deadIdx, liveIdx] : llvm::enumerate(uniformDupeIndexMap)) {
+    if (deadIdx == liveIdx) {
+      continue;
+    }
+    deadArgMap.set(deadIdx);
+    entryBlock.getArgument(deadIdx).replaceAllUsesWith(
+        entryBlock.getArgument(liveIdx));
+  }
+
+  // Update each dispatch site to remove duplicates.
+  SmallVector<unsigned> deadOperands;
+  for (auto idx : deadOperandsMap.set_bits()) {
+    deadOperands.push_back(idx);
+  }
+  for (auto dispatchOp : dispatchOps) {
+    for (auto idx : llvm::reverse(deadOperands)) {
+      dispatchOp.getWorkloadMutable().erase(idx);
+    }
+  }
+
+  // Update the block arguments.
+  entryBlock.eraseArguments(
```

Comment:
you can use `void eraseArguments(const BitVector &eraseIndices);` directly
(it'd be nice if MutableOperandRange also had an erase that took the BitVector too)

Reply:
Thanks, this is great! I updated it here and in the other functions as well.